test/index.spec.js

```diff
@@ -1752,3 +1778,266 @@ describe('api route file analysis', () => {
     )
   })
 })
+
+const middlewareSourceTs = /* typescript */ `
+import { NextResponse } from 'next/server'
+export async function middleware(req: NextRequest) {
+  return NextResponse.next()
+}
+`
+
+const middlewareSourceJs = /* javascript */ `
+import { NextResponse } from 'next/server'
+export async function middleware(req) {
+  return NextResponse.next()
+}
+`
+
+const wait = (seconds = 0.5) => new Promise((resolve) => setTimeout(resolve, seconds * 1000))
+
+const middlewareExists = () => existsSync(resolve('.netlify', 'middleware.js'))
+
+describe('onPreDev', () => {
+  let runtime
+  beforeAll(async () => {
+    runtime = await nextRuntimeFactory({}, { events: new Set(['onPreDev']) })
+  })
+
+  it('should generate the runtime with onPreDev', () => {
+    expect(runtime).toHaveProperty('onPreDev')
+  })
+
+  it('should compile middleware', async () => {
+    await moveNextDist('.next', true)
+    await writeFile(path.join(process.cwd(), 'middleware.ts'), middlewareSourceTs)
+    expect(middlewareExists()).toBeFalsy()
+
+    await runtime.onPreDev(defaultArgs)
+    await wait()
+
+    expect(middlewareExists()).toBeTruthy()
+  })
+})
+
+// skipping for now as the feature works
+// but the tests only seem to run successfully when run locally
+describe('the dev middleware watcher', () => {
+  const watchers = []
+
+  afterEach(async () => {
+    await Promise.all(
+      watchers.map((watcher) => {
+        console.log('closing watcher')
+        return watcher.close()
+      }),
+    )
+    watchers.length = 0
+  })
+
+  it('should compile a middleware file and then exit when killed', async () => {
+    console.log('starting should compile a middleware file and then exit when killed')
+    await moveNextDist('.next', true)
+    await writeFile(path.join(process.cwd(), 'middleware.ts'), middlewareSourceTs)
+    expect(middlewareExists()).toBeFalsy()
+    const { watcher, isReady } = watchForMiddlewareChanges(process.cwd())
+    watchers.push(watcher)
+    await isReady
+    expect(middlewareExists()).toBeTruthy()
+  })
+
+  it.skip('should compile a file if it is written after the watcher starts', async () => {
```

Comment:
I created an issue to reneable the skipped tests, https://github.com/netlify/next-runtime/issues/2005.